update the doc for bin pack scheduling (#33748)
kubernetes/website: content/en/docs/concepts/scheduling-eviction/resource-bin-packing.md

- bsalamat
- k82cn
- ahg-g
title: Resource Bin Packing
content_type: concept
weight: 80
---

<!-- overview -->

> {{< feature-state for_k8s_version="v1.23" state="stable" >}}

Author: Yes. It is useless. I remove it. and  `it is tied to component config version` is mentioned below.

<!-- body -->

## Enabling Bin Packing using MostAllocated
`MostAllocated` allows the users to specify the resources with weight for each resources to 
score nodes on the request to capacity ratio. It favors nodes that have a high allocation of resources.

```yaml
apiVersion: kubescheduler.config.k8s.io/v1beta3
kind: KubeSchedulerConfiguration
profiles:
- pluginConfig:
  - args:
      scoringStrategy:
        resources:
        - name: intel.com/foo
          weight: 3
        - name: intel.com/bar
          weight: 3
        type: MostAllocated
    name: NodeResourcesFit
```

**This feature is disabled by default**

## Enabling Bin Packing using RequestedToCapacityRatio

`RequestedToCapacityRatio` allows the users to specify the resources along with weights for
each resource to score nodes based on the request to capacity ratio. This
allows users to bin pack extended resources by using appropriate parameters
and improves the utilization of scarce resources in large clusters. It favors nodes according to a
configured function of the allocated resources. The
behavior of the `RequestedToCapacityRatio` in `NodeResourcesFit` score function

> can be controlled by a configuration option called `scoringStrategy`. 

Reviewer: ```suggestion
can be controlled by the `scoringStrategy` configuration field.
```

> This argument consists of two parameters `requestedToCapacityRatioParam` and 

Reviewer: ```suggestion
Within the `scoringStrategy` field, you can configure two parameters: `requestedToCapacityRatioParam` and 
```

Reviewer: And maybe you can link to https://kubernetes.io/docs/reference/config-api/kube-scheduler-config.v1beta3/#kubescheduler-config-k8s-io-v1beta3-ScoringStrategy

`resources`. The `shape` in `requestedToCapacityRatioParam` 
parameter allows the user to tune the function as least requested or most 
requested based on `utilization` and `score` values.  The `resources` parameter 
consists of `name` of the resource to be considered during scoring and `weight` 
specify the weight of each resource.

Below is an example configuration that sets
`RequestedToCapacityRatio` to bin packing behavior for extended
resources `intel.com/foo` and `intel.com/bar`.

```yaml
apiVersion: kubescheduler.config.k8s.io/v1beta3
kind: KubeSchedulerConfiguration
profiles:
- pluginConfig:
  - args:
      scoringStrategy:
        resources:
        - name: intel.com/foo
          weight: 3
        - name: intel.com/bar
          weight: 3
        requestedToCapacityRatioParam:
          shape:
          - utilization: 0
            score: 10

>           - utilization: 100

Reviewer: I'm a bit confused here. Shouldn't higher utilization have the higher score? Or is lower score better?

Author: Yes. You are right.  There is an explanation below. But I think it's better to change it.

![image](https://user-images.githubusercontent.com/8958057/168780801-a6eeb495-cf35-49f8-9b73-fcfe19f63bfe.png)
https://kubernetes.io/docs/concepts/scheduling-eviction/resource-bin-packing/

- bsalamat
- k82cn
- ahg-g
title: Resource Bin Packing
content_type: concept
weight: 80
---

<!-- overview -->

{{< feature-state for_k8s_version="v1.23" state="stable" >}}

In the [scheduling-plugin](/docs/reference/scheduling/config/#scheduling-plugins)  `NodeResourcesFit` of kube-scheduler, there are two 
scoring strategies that support the bin packing of resources: `MostAllocated` and `RequestedToCapacityRatio`.

<!-- body -->

> ## Enabling Bin Packing using MostAllocated strategy

Reviewer 2: We should follow the style guide recommendation to use sentence casing for headers. Check out [Style guide section on headers](https://kubernetes.io/docs/contribute/style/style-guide/#headings). Revised header should be "Enabling bin packing using MostAllocated strategy". Same for header further down.

Author: Done

`resources`. The `shape` in `requestedToCapacityRatioParam` 
parameter allows the user to tune the function as least requested or most 
requested based on `utilization` and `score` values.  The `resources` parameter 
consists of `name` of the resource to be considered during scoring and `weight` 
specify the weight of each resource.

Below is an example configuration that sets
`RequestedToCapacityRatio` to bin packing behavior for extended
resources `intel.com/foo` and `intel.com/bar`.

```yaml
apiVersion: kubescheduler.config.k8s.io/v1beta3
kind: KubeSchedulerConfiguration
profiles:
- pluginConfig:
  - args:
      scoringStrategy:
        resources:
        - name: intel.com/foo
          weight: 3
        - name: intel.com/bar
          weight: 3
        requestedToCapacityRatioParam:
          shape:
          - utilization: 0
            score: 0
          - utilization: 100
            score: 10
        type: RequestedToCapacityRatio
    name: NodeResourcesFit
```

Referencing the `KubeSchedulerConfiguration` file with the kube-scheduler 
flag `--config=/path/to/config/file` will pass the configuration to the 
scheduler.

> **You can read the API documentation for [`NodeResourcesFitArgs`](https://kubernetes.io/docs/reference/config-api/kube-scheduler-config.v1beta3/#kubescheduler-config-k8s-io-v1beta3-NodeResourcesFitArgs) to learn about other parameters and their default configuration.**

Reviewer 2: ```suggestion
To learn more about other parameters and their default configuration, see the API documentation for [`NodeResourcesFitArgs`](https://kubernetes.io/docs/reference/config-api/kube-scheduler-config.v1beta3/#kubescheduler-config-k8s-io-v1beta3-NodeResourcesFitArgs)
```

Author: Done

- bsalamat
- k82cn
- ahg-g
title: Resource Bin Packing
content_type: concept
weight: 80
---

<!-- overview -->

In the [scheduling-plugin](/docs/reference/scheduling/config/#scheduling-plugins)  `NodeResourcesFit` of kube-scheduler, there are two 
scoring strategies that support the bin packing of resources: `MostAllocated` and `RequestedToCapacityRatio`.

<!-- body -->

## Enabling bin packing using MostAllocated strategy
The `MostAllocated` strategy scores the nodes based on the utilization of resources, favoring the ones with higher allocation.
For each resource type, you can set a weight to modify its influence in the node score.

> To set the `MostAllocated` strategy for the `NodeResourcesFit` plugin, use a [scheduler configuration](https://kubernetes.io/docs/reference/scheduling/config) similar to the following:

Reviewer 3: Wrap long lines properly; use path rather than full URL for links.

```suggestion
To set the `MostAllocated` strategy for the `NodeResourcesFit` plugin, use a
[scheduler configuration](/docs/reference/scheduling/config) similar to the following:
```

Author: Done

```yaml
apiVersion: kubescheduler.config.k8s.io/v1beta3
kind: KubeSchedulerConfiguration
profiles:
- pluginConfig:
  - args:
      scoringStrategy:
        resources:
        - name: cpu
          weight: 1
        - name: memory
          weight: 1
        - name: intel.com/foo
          weight: 3
        - name: intel.com/bar
          weight: 3
        type: MostAllocated
    name: NodeResourcesFit
```

> To learn more about other parameters and their default configuration, see the API documentation for [`NodeResourcesFitArgs`](https://kubernetes.io/docs/reference/config-api/kube-scheduler-config.v1beta3/#kubescheduler-config-k8s-io-v1beta3-NodeResourcesFitArgs).

Reviewer 3: Same here.

```suggestion
To learn more about other parameters and their default configuration, see the API documentation for 
[`NodeResourcesFitArgs`](/docs/reference/config-api/kube-scheduler-config.v1beta3/#kubescheduler-config-k8s-io-v1beta3-NodeResourcesFitArgs).
```

Author: Done

> To learn more about other parameters and their default configuration, see the API documentation for [`NodeResourcesFitArgs`](https://kubernetes.io/docs/reference/config-api/kube-scheduler-config.v1beta3/#kubescheduler-config-k8s-io-v1beta3-NodeResourcesFitArgs) to learn about other parameters and their default configuration.

Reviewer: you put "to learn more" twice 😅

Author: 。。。 sorry Updated

## Enabling bin packing using RequestedToCapacityRatio

The `RequestedToCapacityRatio` strategy allows the users to specify the resources along with weights for
each resource to score nodes based on the request to capacity ratio. This
allows users to bin pack extended resources by using appropriate parameters
to improve the utilization of scarce resources in large clusters. It favors nodes according to a

> configured function of the allocated resources. The
> behavior of the `RequestedToCapacityRatio` in `NodeResourcesFit` score function
> can be controlled by the [scoringStrategy](/docs/reference/config-api/kube-scheduler-config.v1beta3/#kubescheduler-config-k8s-io-v1beta3-ScoringStrategy)  configuration field.

Reviewer 3: Line wrappings here can be adjusted:

```suggestion
configured function of the allocated resources. The behavior of the `RequestedToCapacityRatio` in 
the `NodeResourcesFit` score function can be controlled by the
[scoringStrategy](/docs/reference/config-api/kube-scheduler-config.v1beta3/#kubescheduler-config-k8s-io-v1beta3-ScoringStrategy)
field.
```

Author: Done

Within the `scoringStrategy` field, you can configure two parameters: `requestedToCapacityRatioParam` and
`resources`. The `shape` in `requestedToCapacityRatioParam` 
parameter allows the user to tune the function as least requested or most 
requested based on `utilization` and `score` values.  The `resources` parameter 
consists of `name` of the resource to be considered during scoring and `weight` 
specify the weight of each resource.

Below is an example configuration that sets

> `RequestedToCapacityRatio` to bin packing behavior for extended
> resources `intel.com/foo` and `intel.com/bar`.

Reviewer 3: ```suggestion
the bin packing behavior for extended resources `intel.com/foo` and `intel.com/bar`
using the `requestedToCapacityRatio` field.
```

Author: Done

```yaml
apiVersion: kubescheduler.config.k8s.io/v1beta3
kind: KubeSchedulerConfiguration
profiles:
- pluginConfig:
  - args:
      scoringStrategy:
        resources:
        - name: intel.com/foo
          weight: 3
        - name: intel.com/bar
          weight: 3
        requestedToCapacityRatioParam:
          shape:
          - utilization: 0
            score: 0
          - utilization: 100
            score: 10
        type: RequestedToCapacityRatio
    name: NodeResourcesFit
```

Referencing the `KubeSchedulerConfiguration` file with the kube-scheduler 
flag `--config=/path/to/config/file` will pass the configuration to the 
scheduler.

> To learn more about other parameters and their default configuration, see the API documentation for [`NodeResourcesFitArgs`](https://kubernetes.io/docs/reference/config-api/kube-scheduler-config.v1beta3/#kubescheduler-config-k8s-io-v1beta3-NodeResourcesFitArgs).

Reviewer 3: ```suggestion
To learn more about other parameters and their default configuration, see the API documentation for 
[`NodeResourcesFitArgs`](/docs/reference/config-api/kube-scheduler-config.v1beta3/#kubescheduler-config-k8s-io-v1beta3-NodeResourcesFitArgs).
```

Author: Done

### Tuning the score function

`shape` is used to specify the behavior of the

> `RequestedToCapacityRatio` function.

Reviewer 3: merge line 97 to 96?
